doc/csync.1
'\" t
.\"     Title: csync
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 04/03/2013
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "CSYNC" "1" "04/03/2013" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
csync \- ownCloud CLI client written in python
.SH "SYNOPSIS"
.sp
\fBcsync\fR [\-h] [\-v] [\-c \fICONFIG\fR] [\-u \fIUSER\fR] [\-\-ssl \fISSLFINGERPRINT\fR] [\-p \fIPASS\fR] [\-\-dry\-run] [\-\-debug] [\-s \fISRC\fR] [\-d \fIDST\fR] [\-\-url \fIURL\fR] [\-\-use\-keyring]
.SH "DESCRIPTION"
.sp
Syncing files across machines using ownCloud DAV server\&.
.PP
\fBPassword options:\fR
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
You can specify on the cmd line:
\fB\-p\fR
(not very safe)
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
In the environment variable:
\fIOCPASS\fR
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
In the owncloud\&.cfg file as
\fBpass = \fR\fB\fI<password>\fR\fR
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Use keyring to store passwords in a keyring (needs Python Keyring Lib to be installed)
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Do none of the above, and it will prompt you for the password
.RE
.sp
The choice is yours, if you put it in the cfg file, be careful to make sure nobody but you can read the file\&. (0400/0600 file perms)\&.
.SH "OPTIONS"
.PP
\fB\-c \fR\fB\fICONFIG\fR\fR\fB, \-\-config \fR\fB\fICONFIG\fR\fR
.RS 4
Configuration file to use
.RE
.PP
\fB\-u \fR\fB\fIUSER\fR\fR\fB, \-\-user \fR\fB\fIUSER\fR\fR
.RS 4
Username on the ownCloud server
.RE
.PP
\fB\-p \fR\fB\fIPASS\fR\fR\fB, \-\-pass \fR\fB\fIPASS\fR\fR
.RS 4
Password on the ownCloud server ( You can also store this in environment variable
\fIOCPASS\fR
)
.RE
.PP
\fB\-s \fR\fB\fISRC\fR\fR\fB, \-\-src \fR\fB\fISRC\fR\fR
.RS 4
Local directory to sync against
.RE
.PP
\fB\-d \fR\fB\fIDST\fR\fR\fB, \-\-dst \fR\fB\fIDST\fR\fR
.RS 4
Folder on server to sync against
.RE
.PP
\fB\-\-url \fR\fB\fIURL\fR\fR
.RS 4
URL of the ownCloud server
.RE
.PP
\fB\-\-ssl \fR\fB\fISSLFINGERPRINT\fR\fR
.RS 4
valid SSL fingerprint for the ownCloud server
.RE
.PP
\fB\-\-use\-keyring\fR
.RS 4
use keyring if available to store password safely (Needs Python Keyring Lib to be installed)
.RE
.PP
\fB\-\-dry\-run\fR
.RS 4
Dry Run, do not actually execute command
.RE
.PP
\fB\-\-debug\fR
.RS 4
Print a bunch of debug info
.RE
.PP
\fB\-h, \-\-help\fR
.RS 4
show help message and exit
.RE
.PP
\fB\-v, \-\-version\fR
.RS 4
show program\(cqs version number and exit
.RE
.SH "CONFIGURATION FILE"
.sp
The csync configuration file can be found in \fI$HOME/\&.local/share/data/ownCloud/owncloud\&.cfg\fR\&. Only the \fBownCloud\fR section is supported\&. This file is also ued by the normal owncloud\-client\&.
.sp
Supported keys in the cfg file:
.PP
\fBuser\fR
.RS 4
username on the ownCloud server
.RE
.PP
\fBpass\fR
.RS 4
password on the ownCloud server
.RE
.PP
\fBurl\fR
.RS 4
url of the ownCloud server
.RE
.PP
\fBsslfingerprint\fR
.RS 4
valid SSL fingerprint for the ownCloud server
.RE
.PP
\fBsrc\fR
.RS 4
local directory to sync against
.RE
.PP
\fBdst\fR
.RS 4
folder on the server to sync against
.RE
.sp
A complete example:
.sp
.if n \{\
.RS 4
.\}
.nf
[ownCloud]
user=awesomeSauce
pass=PasswordThisIsSuperSuperSecretReallyISwearLOL
url=https://www\&.example\&.org/owncloud/
sslfingerprint=
src=/home/awesomeSauce/ownCloud
dst=clientsync
.fi
.if n \{\
.RE
.\}
.SH "BUGS"
.sp
Please report bugs at https://github\&.com/csawyerYumaed/pyOwnCloud/issues\&.
.SH "SEE ALSO"
.sp
\fBlibcsync\fR(7)
.SH "AUTHORS"
.sp
pyOwnCloud Team <pyowncloud@lists\&.cknow\&.org>
.SH "COPYING"
.sp
Copyright (c) 2013 pyOwnCloud Team\&. Free use of this software is granted under the terms of the GNU General Public License 2 (GPL\-2)\&.


CHANGELOG
CHANGELOG
=========

version 0.3 (unreleased - this updated to 20655515dedf8495f2701ec045b6b5c5e0dad25c [2065551])
	* binary command is now called oclient.
    * will prompt for password if not specified in other ways.
	* has a setup.py and can be properly 'packaged' now.  
	* better finding of the csync C library
	* uses python-keyring to use the same passwords as the owncloud-client if
	available
	* LICENSE and copyright changes, now all GPL-2
	* fixed how we handle versioning to be better.


version 0.2 (releaded 2013-03-12)
-----------
	* Lots of UI cleanup
	* loads more error checking
	* Works beyond the orignal author's machine.
	* Thanks to Weuxel (https://github.com/Weuxel) for lots of documenation
	and cleanup.
	* Thanks to duck (https://github.com/duck) for lots of testing!

version 0.1  (released 2013-03-09)
-----------
	* Initial release


csync/csync.py
#!/usr/bin/env python

import os
import sys
import argparse
import ConfigParser
import ctypes
import re
import pprint
import copy
import getpass

try:
	import keyring
except:
	keyring = None

import csynclib
import version
VERSION = version.version

#Use global variables for user/pass & fingerprint because we have to handle this C callback stuff.
USERNAME = ''
PASSWORD = ''
PASSWORD_SAFE = '********'
SSLFINGERPRINT = ''
DEBUG = False
USE_KEYRING = False

def authCallback(prompt, buffer, bufferLength, echo, verify, userData):
	"""
	(const char *prompt, char *buf, size_t len,
		int echo, int verify, void *userdata)
	called like this:
		("Enter your username: ", buf, NE_ABUFSIZ-1, 1, 0, dav_session.userdata )
		type is 1 for username, 0 for password.
	"""
	if DEBUG:
		print 'authCallback:', prompt,  buffer,  bufferLength, echo, verify, userData
		#print 'string:', ctypes.string_at(buffer, bufferLength-1)
	ret = None
	if 'username' in prompt:
		ret = USERNAME
	elif 'password' in prompt:
		if keyring and USE_KEYRING:
			print "using password from keyring"
			ret = keyring.get_password('ownCloud', USERNAME)
		if ret is None:
			if not PASSWORD:
				ret = getpass.getpass('ownCloud password:')
			else:
				ret = PASSWORD
			if keyring and USE_KEYRING:
				print "saving password to keyring"
				keyring.set_password('ownCloud', USERNAME, ret)
	elif 'SSL' in prompt:
		fingerprint = re.search("fingerprint: ([\\w\\d:]+)", prompt).group(1)
		if fingerprint == SSLFINGERPRINT:
			ret = 'yes'
		else:
			print 'SSL fingerprint: %s not accepted, aborting' % fingerprint
			ret = 'no'
	else:
		print 'authCallback: unknown prompt:', prompt
		return -1
	bufferLength = len(ret)
	for i in range(len(ret)):
		ctypes.memset(buffer+i, ord(ret[i]), 1)
	if DEBUG:
		buffString = ctypes.string_at(buffer, bufferLength)
		if PASSWORD:
			if PASSWORD in buffString:
				buffString = buffString.replace(PASSWORD, PASSWORD_SAFE)
		print 'returning:', buffString
	return 0

class ownCloudSync():
	"""This handles the actual syncying with ownCloud
	cfg is a {}.  should have these things:
		user:
		pass:
		url:
		src:
	None of them are optional. :)
	optional items:
		SSLfingerPrint:
	"""
	def __init__(self, cfg = None):
		"""initialize"""
		self.cfg = cfg
		global USERNAME, PASSWORD, SSLFINGERPRINT, USE_KEYRING
		USERNAME = cfg['user']
		PASSWORD = cfg['pass']
		SSLFINGERPRINT = cfg['sslfingerprint']
		USE_KEYRING = cfg['use_keyring']
		libVersion = csynclib.csync_version(0,40,1)
		if DEBUG:
			print 'libocsync version: ', libVersion
		if libVersion != '0.70.4':
			print 'This version of libocsync %s is not tested against ownCloud server 4.7.5.' % libVersion
		c = csynclib.CSYNC()
		self.ctx = ctypes.pointer(c)
		self.buildURL()
		#pprint.pprint(self.cfg)
		print 'Syncing %s to %s logging in as user: %s' %  (self.cfg['src'], 
			self.cfg['url'],
			USERNAME,
			)
		if cfg.has_key('dry_run'):
			return
		self.sync()

	def buildURL(self):
		"""build the URL we use for owncloud"""
		url = self.cfg['url']
		if url == '':
			print 'You must specify a url, use --url, or put in cfg file.'
			sys.exit(1j)
		url = url.replace('https','ownclouds')
		url = url.replace('http','owncloud')
		#add / if needed
		if url[-1:] != '/':
			url = ''.join((url,'/'))
		url += self.cfg['davPath']
		#add / if needed
		if url[-1:] != '/':
			url = ''.join((url,'/'))
		url = ''.join((url, self.cfg['dst']))
		#take off any trailing slash.
		if url[-1:] == '/':
			url = url[:-1]
		self.cfg['url'] = url
		if DEBUG:
			print 'buildURL: ', url
		return



	def sync(self):
		r = csynclib.csync_create(self.ctx, self.cfg['src'], self.cfg['url'])
		if r != 0:
			error(self.ctx,'csync_create', r)
		csynclib.csync_set_log_callback(self.ctx, csynclib.csync_log_callback(log))
		acb = csynclib.csync_auth_callback(authCallback)
		if DEBUG:
			print 'authCallback setup'
		csynclib.csync_set_auth_callback(self.ctx, acb)

		r = csynclib.csync_init(self.ctx)
		if r != 0:
			error(self.ctx, 'csync_init', r)
		if DEBUG:
			print 'Initialization done.'
		#csynclib.csync_set_log_verbosity(self.ctx, ctypes.c_int(11))
		r = csynclib.csync_update(self.ctx)
		if r != 0:
			error(self.ctx, 'csync_update', r)
		if DEBUG:
			print 'Update done.'
		r = csynclib.csync_reconcile(self.ctx)
		if r != 0:
			error(self.ctx, 'csync_reconcile', r)
		if DEBUG:
			print 'Reconcile done.'
		r = csynclib.csync_propagate(self.ctx)
		if r != 0:
			error(self.ctx, 'csync_propogate', r)
		if DEBUG:
			print 'Propogate finished, destroying.'
		r = csynclib.csync_destroy(self.ctx)
		if r != 0:
			error(self.ctx, 'csync_destroy', r)
		return

def log(ctx, verbosity, function, buffer, userdata):
	"""Log stuff from the ocsync library, but it does not work..."""
	print 'LOG:', verbosity, function, buffer, userdata
	return 0

def error(ctx, cmd, returnCode):
	"""handle library errors"""
	errNum = csynclib.csync_get_error(ctx)
	errMsg = csynclib.csync_get_error_string(ctx)
	if not errMsg:
		if errNum == 20 and cmd == 'csync_update':
			errMsg = 'This is an authentication problem with the server, check user/pass.'
		if errNum == 26 and cmd == 'csync_update':
			errMsg = 'This is a remote folder destination issue, check that the remote folder exists on ownCloud.'
	print 'ERROR: %s exited %s, error %s: %s' % (
		cmd,
		returnCode,
		errNum,
		errMsg,
		)
	sys.exit(1)


def getConfigPath():
	"""get the local configuration file path
	"""
	if sys.platform.startswith('linux'):
		cfgPath = os.path.join('~','.local','share','data','ownCloud')
		cfgPath = os.path.expanduser(cfgPath) 
	elif sys.platform == 'darwin':
		cfgPath = os.path.join('~','Library','Application Support','ownCloud')
		cfgPath = os.path.expanduser(cfgPath) 
	elif 'win' in sys.platform:
		cfgPath = os.path.join('%LOCALAPPDATA%','ownCloud')
		cfgPath = os.path.expandvars(cfgPath)
	else:
		print 'Unkown/not supported platform %s, please file a bug report. ' % sys.platform
		sys.exit(1)
	if DEBUG:
		print 'getConfigPath:', cfgPath
	return cfgPath

def getConfig(args):
	if DEBUG:
		print 'From args: '
		pargs = copy.copy(args)
		if pargs['pass']:
			pargs['pass'] = PASSWORD_SAFE
		pprint.pprint(pargs)
	newArgs = {}
	for k, v in args.iteritems():
		if v:
			newArgs[k] = v
	args = newArgs
	cfg = {}
	cfgFile = None
	if args.has_key('config'):
		cfgFile = args['config']
	else:
		cfgPath = getConfigPath()
		if os.path.exists(os.path.join(cfgPath,'owncloud.cfg')):
			cfgFile = os.path.join(cfgPath, 'owncloud.cfg')
	if cfgFile:
		with open(cfgFile) as fd:
			"""We use the INI file format that Mirall does. we allow more
			things in the cfg file...
				pass: the password
			"""
			c = ConfigParser.SafeConfigParser()
			c.readfp(fd)
			cfg = dict(c.items('ownCloud'))
			if DEBUG:
				print 'conifguration info received from %s:' % cfgFile
				pcfg = copy.copy(cfg)
				if pcfg.has_key('pass'):
					pcfg['pass'] = PASSWORD_SAFE
				pprint.pprint(pcfg)
	cfg.setdefault('davPath', 'remote.php/webdav/')
	cfg.setdefault('sslfingerprint' '')
	cfg.setdefault('pass', None)
	cfg.setdefault('user', getpass.getuser())
	cfg.setdefault('use_keyring', False)
	if os.environ.has_key('OCPASS'):
		cfg['pass'] = os.environ['OCPASS']
		if DEBUG:
			print 'password coming from environment'
	#cmd line arguments win out over config files.
	cfg.update(args)
	if DEBUG:
		print 'Finished config file:'
		pcfg = copy.copy(cfg)
		if pcfg.has_key('pass'):
			pcfg['pass'] = PASSWORD_SAFE
		pprint.pprint(pcfg)
	return cfg

def startSync(args):
	if args['debug']:
		global DEBUG
		DEBUG = True
		print 'Turning debug on'
	cfg = getConfig(args)
	try:
		sync = ownCloudSync(cfg)
	except KeyError:
		exc_type, exc_value, exc_tb = sys.exc_info()
		print 'Sorry this option: %s is required, was not found in cfg file or on cmd line.' % (exc_value)
		if DEBUG:
			raise

def main():
	parser = argparse.ArgumentParser(
		formatter_class=argparse.RawDescriptionHelpFormatter,
		description = 'Synchronize files across machines using ownCloud DAV server.',
		epilog = """
I support the ownCloud config file, which is located here:
  {cfg}
I only support the 'ownCloud' section of the config.
I support the following keys in the cfg  file:
	user: The username on the ownCloud server
	url: the url of the ownCloud Server
	pass: the password on the ownCloud server
	sslfingerprint: a valid SSL fingerprint for the server.
	src: local directory to sync against.
	dst: folder on the server to sync against.
complete example:
[ownCloud]
user=awesomeSauce
pass=PasswordThisIsSuperSuperSecretReallyISwearLOL
url=https://www.example.org/owncloud/
sslfingerprint=
src=/home/awesomeSauce/ownCloud
dst=clientsync

Password options:
  *) You can specify on the cmd line: -p (not very safe)
  *) In the envifonment variable: OCPASS
  *) In the owncloud.cfg file as pass = <password>
  *) Do none of the above, and it will prompt you for the password.
  *) Use keyring to store passwords in a keyring. (keyring package is {keyring}installed).
  The choice is yours, if you put it in the cfg file, be careful to 
  make sure nobody but you can read the file. (0400/0600 file perms).
		""".format(cfg = os.path.join(getConfigPath(),'owncloud.cfg'), keyring="" if keyring else "NOT "),
	)
	v = "%s - repo: %s" % (VERSION.asString, VERSION.asHead)
	parser.add_argument('-v', '--version', 
		action='version', 
		version = '%(prog)s ' + v)
	parser.add_argument('-c', '--config', nargs='?', default = None,
		help = "Configuration to use.")
	parser.add_argument('-u', '--user', nargs='?', default = None,
		help = "Username on server.")
	parser.add_argument('--ssl', nargs='?', default = None,
		dest = 'sslfingerprint',
		help = "SSL fingerprint on server to accept.")
	parser.add_argument('-p', '--pass', nargs='?', default = None,
		help = "Password on server. You can also store this in environment variable OCPASS.")
	parser.add_argument('--dry-run', action = 'store_true', default = False,
		help = "Dry Run, do not actually execute command.")
	parser.add_argument('--debug', action = 'store_true', default = False,
		help = "Print a bunch of debug info.")
	parser.add_argument('-s', '--src', nargs='?',
		default =  os.path.expanduser(os.path.join('~','ownCloud')),
		 help = "Local Directory to sync with.")
	parser.add_argument('-d', '--dst', nargs='?', default = 'clientsync',
		help = "Folder on server.")
	parser.add_argument('--url', nargs='?', default = None,
		 help = "URL to sync to.")
	if keyring:
		parser.add_argument('--use-keyring', action = 'store_true', default = False,
				help = "use keyring if available to store password safely.")
	args = vars(parser.parse_args())
	startSync(args)

if __name__ == '__main__':
	main()
